network_simulator/link.py

```diff
@@ -4,44 +4,98 @@
 """
 
 from Queue import Queue
+from packet import *
+from common import *
+from event import *
+import logging
+
+logger = logging.getLogger(__name__)
 
-# TODO(yubo): Finish implementing the classes and Events for this.
 
 
 class LinkBuffer(object):
     """
     Representation of a Link's FIFO buffer.
-
-    TODO(team): Parameters related to link congestion, so we can have dynamic
-    routing table updates.
-    TODO(yubo): Functions for whether this is full etc.
     """
 
-    def __init__(self, buffer_size_bits=None, queue=Queue()):
+    # TODO(team): Parameters related to link congestion, so we can have dynamic
+    # routing table updates.
+    # TODO(yubo): packet_exemption_times deque
+
+    def __init__(self, max_buffer_size_bits=None, queue=Queue()):
         """
-        :ivar int buffer_size_bits: maximum buffer size in bits.
+        :ivar int max_buffer_size_bits: maximum buffer size in bits.
         :ivar Queue queue: a FIFO queue of LinkBufferElements
+        :ivar int curr_buffer_size_bits: current buffer size in bits
         """
-        self.buffer_size_bits = buffer_size_bits
+        self.max_buffer_size_bits = max_buffer_size_bits
         self.queue = queue
 
     def __repr__(self):
         return str(self.__dict__)
 
+    def remaining(self):
+        """
+        returns remaining size in buffer in bits
+
+        not strictly necessary, push() will check remaining() condition
+        :return: remaining buffer size in bits
+        """
+        return self.max_buffer_size_bits - self.curr_buffer_size_bits
+
+    def push(self, packet, global_clock_sec):
+        """
+        Pushes packet onto LinkBuffer going towards packet.dest_id
+
+        :param Packet packet: Packet to push
+        :return: True if pushed, False if not
+        """
+        if isinstance(packet, RouterPacket):
+            # always push RouterPacket
+            self.queue.put(LinkBufferElement(\
```

Comment:
LinkBufferElement needs to include destination Device as mentioned below. Also, since the code for the LinkBufferElement you construct doesn't depend on the packet type, just write one line before the if instead of multiple.